x-pack/plugins/reporting/server/routes/generate_from_savedobject.ts

```diff
@@ -0,0 +1,157 @@
+ *     - query bar
+ *     - local (transient) changes the user made to the saved object
+ */
+export function registerGenerateCsvFromSavedObject(
+  server: KbnServer,
+  handleRoute: HandlerFunction,
+  handleRouteError: HandlerErrorFunction
+) {
+  const getRouteConfig = getRouteConfigFactoryReportingPre(server);
+  const routeOptions = {
+    ...getRouteConfig(() => CSV_FROM_SAVEDOBJECT_JOB_TYPE),
+    validate: {
+      params: Joi.object({
+        savedObjectType: Joi.string().required(),
+        savedObjectId: Joi.string().required(),
+      }).required(),
+      payload: Joi.object({
+        state: Joi.object().default({}),
+        timerange: Joi.object({
+          timezone: Joi.string().default('UTC'),
+          min: Joi.date().required(),
+          max: Joi.date().required(),
+        }).optional(),
+      }),
+    },
+  };
+  const getDocumentPayload = getDocumentPayloadFactory(server);
+
+  // csv: immediate download
+  server.route({
+    path: `${BASE_GENERATE}/immediate/csv/saved-object/{savedObjectType}:{savedObjectId}`,
+    method: 'POST',
+    options: routeOptions,
+    handler: async (request: Request, h: ResponseToolkit) => {
+      /*
+       * 1. Queue a job with getJobFromRouteHandler
+       *   - `isImmediate: true` gets us the complete result data in payload.objects
+       * 2. Copy the completed data stashed in the job as output.content
+       *   - Makes the content available for download
+       * 3. Return a response with CSV content
+       */
+      const queuedJob: QueuedJobPayload = await getJobFromRouteHandler(
+        handleRoute,
+        handleRouteError,
+        request,
+        h,
+        {
+          isImmediate: true,
+        }
+      );
+
+      // FIXME this is REALLY ugly
+      const jobSource: JobDoc = get(queuedJob, 'source.job');
+      const output: JobDocOutput = getDocumentPayload({
+        _source: {
+          ...jobSource,
+          status: 'completed',
+          output: {
+            content: jobSource.payload.objects,
+            content_type: 'text/csv',
+          },
+        },
+      });
+
+      const response: ResponseObject = h
+        .response(output.content)
+        .type(output.contentType)
+        .code(output.statusCode);
+
+      if (output.headers) {
+        Object.keys(output.headers).forEach(key => {
+          response.header(key, output.headers[key]);
+        });
+      }
+
+      const { isBoom } = response as KibanaResponse;
+      if (isBoom == null) {
+        response.header('accept-ranges', 'none');
+      }
+
+      return response;
+    },
+  });
+
+  // csv: queue job for execution
+  server.route({
+    path: `${BASE_GENERATE}/csv/saved-object/{savedObjectType}:{savedObjectId}`,
+    method: 'POST',
+    options: routeOptions,
+    handler: async (request: Request, h: ResponseToolkit) => {
+      return getJobFromRouteHandler(handleRoute, handleRouteError, request, h, {
+        isImmediate: false,
+      });
+    },
+  });
```

Comment:
This file is in need of re-organization. I could see a one-route-per file convention for this: the `getJobFromRouteHandler ` helper code no longer applies to the `immediate` route